File: algorithms/8.c
#include <stdio.h>
#include <string.h>

int myAtoi(char* str)
{
    if ( !str || str[0]=='\0' )
        return 0;
    int i = 0;
    int sign = 0, overflow = 0;
    unsigned int result = 0;
    while ( str[i] != '\0' && str[i] == ' ' )
        i++;
    if ( str[i] != '\0' && (str[i] == '+'||str[i] == '-') )
        if ( str[i++] == '-' )
            sign = 1;
    while ( str[i] != '\0' )
    {
        if ( str[i] >= '0' && str[i] <= '9' )
        {
            if ( result > 4294967296/10 )
            {
                overflow = 1;
                break;
            }
            result = result*10+str[i]-'0';
        }
        else
            break;
        i++;
    }
    if ( sign )
    {
        if ( overflow || result > 2147483648 )
            return -2147483648;
        else
            return -(int)result;
    }
    else
    {
        if ( overflow || result > 2147483647 )
            return 2147483647;
        else
            return (int)result;
    }
}

int main()
{
    printf("%d\n", myAtoi("2147483647"));
    return 0;
}

File: algorithms/python3/8.py
class Solution:
    def myAtoi(self, s):
        """
        :type s: str
        :rtype: int
        """
        s = s.strip()
        if s == '':
            return 0

        i, sign = 0, 1
        if s[i] == '+' or s[i] == '-':
            sign = -1 if s[i] == '-' else 1
            i += 1

        rls = 0
        while i < len(s) and s[i].isdigit():
            rls = rls * 10 + int(s[i])
            i += 1

        rls = rls * sign
        rls = min(rls, +2147483647)
        rls = max(rls, -2147483648)
        return rls
